### PR TITLE
ADM 588: Export 7 Key metrics in csv report

### PR DESCRIPTION
**Pull request by TingyuDong**
## Summary

Currently, the page displays data for 7 metrics but does not allow exporting the related CSV files. Therefore, we need to add an `Export metric data` button on the report page. Upon clicking this button, a CSV file containing data for the 7 metrics can be downloaded.

## Before

Currently the 7 key metrics could only be viewed within report page.

image

## After

When clicking on export key metrics button, we will download the key metrics report in csv format

image

image

## Note

_Null_

### CODE REVIEW

File: backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java [hunk 1]
```diff
@@ -260,4 +261,101 @@ public void shouldHasContentWhenGetDataFromCsvGivenDataTypeIsBoard() throws IOEx
 		file.delete();
 	}
 
+	@Test
+	void shouldConvertMetricDataToCsv() throws IOException {
+		ReportResponse reportResponse = MetricCsvFixture.MOCK_METRIC_CSV_DATA();
+		csvFileGenerator.convertMetricDataToCSV(reportResponse, mockTimeStamp);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java:267]: given when then, and `new line` before each

File: backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java [hunk 2]
```diff
@@ -260,4 +261,101 @@ public void shouldHasContentWhenGetDataFromCsvGivenDataTypeIsBoard() throws IOEx
...
+		assertThrows(FileIOException.class, () -> csvFileGenerator.getDataFromCSV("metric", 1686710104536L));
+		assertThrows(FileIOException.class,
+				() -> csvFileGenerator.convertMetricDataToCSV(reportResponse, "15469:89/033"));
+	}
+
+	@Test
+	public void shouldHasContentWhenGetDataFromCsvGivenDataTypeIsMetric() throws IOException {
+		ReportResponse reportResponse = MetricCsvFixture.MOCK_METRIC_CSV_DATA();
+
+		csvFileGenerator.convertMetricDataToCSV(reportResponse, mockTimeStamp);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/report/CSVFileGeneratorTest.java:312]: new line after `when`

File: backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java [hunk 3]
```diff
@@ -110,6 +132,7 @@ public void convertPipelineDataToCSV(List<PipelineCSVInfo> leadTimeData, String
 
 	public InputStreamResource getDataFromCSV(String dataType, long csvTimeStamp) {
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:133]: csvTuneStamp is not readable for anyone, can u help to refactor it with a readable value, like `2023-02-02-14-14-15`

File: backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java [hunk 4]
```diff
@@ -110,6 +132,7 @@ public void convertPipelineDataToCSV(List<PipelineCSVInfo> leadTimeData, String
...
 		return switch (dataType) {
+			case "metric" -> readStringFromCsvFile(CSVFileNameEnum.METRIC.getValue() + "-" + csvTimeStamp + ".csv");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:135]: extract `.csv` as a static constant, and `-`

File: frontend/__tests__/src/components/Metrics/ReportStep/ReportStep.test.tsx [hunk 5]
```diff
@@ -217,4 +218,12 @@ describe('Report Step', () => {
       expect(getByTestId('loading-page')).toBeInTheDocument()
     })
   })
+
+  it('should show export metric button when visiting this page', async () => {
+    const { queryByText } = await act(() => setup(['']))
+
+    const exportMetricButton = queryByText(EXPORT_METRIC_DATA)
+
```
Comment (guzhongren) [inline frontend/__tests__/src/components/Metrics/ReportStep/ReportStep.test.tsx:226]: not `new line` on line 226

**Review by guzhongren — COMMENTED** [review]
just some small changes, thx

File: backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java (hunk 3 above)
Comment (TingyuDong) [reply backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:133]: This `csvTuneStamp` is only used for naming the CSV files stored on the server. The filename for the CSV files that our users download is defined on the frontend, and the timestamp has already been formatted as 'HHmmSSS', such as metric-20231009-20231020-1947776.csv. Would you think this format is acceptable?

Comment (guzhongren) [reply backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:137]: No, storing it as a readable name is a basic requirement everywhere, coc, the code and the output are read/identified by a human, not the robot

Comment (TingyuDong) [reply backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:137]: You're right, I've learned. So, do you think it's better for us to format this timestamp in a GET request on the frontend or to format it on the backend?

Comment (TingyuDong) [reply backend/src/main/java/heartbeat/service/report/CSVFileGenerator.java:137]: You're right, I've learned. So, do you think it's better for us to format this timestamp on the frontend or to format it on the backend?

**Review by davidwangcn9 — APPROVED** [review]
LGTM

**Pull request merged**